addons/http_routing/models/ir_http.py

```diff
@@ -25,308 +24,127 @@
+    def to_python(self, value) -> models.AbstractModel:
+        record = super().to_python(value)
+        if record.id < 0 and not record.browse(record.id).exists():
             # limited support for negative IDs due to our slug pattern, assume abs() if not found
-            if not env[self.model].browse(record_id).exists():
-                record_id = abs(record_id)
-        return env[self.model].with_context(_converter_value=value).browse(record_id)
+            record = record.browse(abs(record.id))
+        return record.with_context(_converter_value=value)
 
 
 class IrHttp(models.AbstractModel):
     _inherit = ['ir.http']
 
     rerouting_limit = 10
 
+    # ------------------------------------------------------------
+    # Slug tools
+    # ------------------------------------------------------------
+
+    @classmethod
+    def _slugify_one(cls, value: str, max_length: int = 0) -> str:
+        """ Transform a string to a slug that can be used in a url path.
+            This method will first try to do the job with python-slugify if present.
+            Otherwise it will process string by stripping leading and ending spaces,
+            converting unicode chars to ascii, lowering all chars and replacing spaces
+            and underscore with hyphen "-".
+            :param s: str
+            :param max_length: int
+            :rtype: str
+        """
+        value = ustr(value)
+        if slugify_lib:
+            # There are 2 different libraries only python-slugify is supported
+            try:
+                return slugify_lib.slugify(value, max_length=max_length)
+            except TypeError:
+                pass
+        uni = unicodedata.normalize('NFKD', value).encode('ascii', 'ignore').decode('ascii')
+        slug_str = re.sub(r'[\W_]', ' ', uni).strip().lower()
+        slug_str = re.sub(r'[-\s]+', '-', slug_str)
+        return slug_str[:max_length] if max_length > 0 else slug_str
+
+    @classmethod
+    def _slugify(cls, value: str, max_length: int = 0, path: bool = False) -> str:
+        if not path:
+            return cls._slugify_one(value, max_length=max_length)
+        else:
+            res = []
+            for u in value.split('/'):
+                s = cls._slugify_one(u, max_length=max_length)
+                if s:
+                    res.append(s)
+            # check if supported extension
+            path_no_ext, ext = os.path.splitext(value)
+            if ext in EXTENSION_TO_WEB_MIMETYPES:
+                res[-1] = cls._slugify_one(path_no_ext) + ext
+            return '/'.join(res)
+
+    @classmethod
+    def _slug(cls, value: models.BaseModel | str) -> str:
+        try:
+            if not value.id:
+                raise ValueError("Cannot slug non-existent record %s" % value)
+            name = value.display_name
+            slugname = name and cls._slugify(name).strip().strip('-')
+            return f"{slugname}-{value.id}" if slugname else str(value.id)
```

Comment:
if we are to rewrite it
```suggestion
            if not value.display_name:
                return str(value.id)
            slugname = cls._slugify(value.display_name).strip().strip('-')
            return f"{slugname}-{value.id}" if slugname else str(value.id)
```